src/cli/proxy.rs

```diff
@@ -638,29 +699,19 @@ mod tests {
         config.clusters.modify(|clusters| {
             clusters.insert_default(
                 [crate::net::endpoint::Endpoint::from(
-                    endpoint.socket.local_ipv4_addr().unwrap(),
+                    endpoint.socket.local_addr().unwrap(),
                 )]
                 .into(),
             )
         });
 
-        let shared_socket = Arc::new(
-            DualStackLocalSocket::new(
-                crate::test::available_addr(&AddressType::Random)
-                    .await
-                    .port(),
-            )
-            .unwrap(),
-        );
-        let sessions = SessionPool::new(
-            config.clone(),
-            shared_socket.clone(),
-            tokio::sync::watch::channel(()).1,
-        );
+        let (tx, rx) = async_channel::unbounded();
+        let (_shutdown_tx, shutdown_rx) = tokio::sync::watch::channel(());
 
-        proxy
-            .run_recv_from(&config, &sessions, shared_socket)
-            .unwrap();
+        let sessions = SessionPool::new(config.clone(), tx, shutdown_rx);
+
+        proxy.run_recv_from(&config, &sessions, rx).unwrap();
+        tokio::time::sleep(Duration::from_millis(500)).await;
```

Comment:
My suggestion here (if possible) would be to create a looping poll to check a condition (or run the whole test), with a sleep at the end of each loop, and break the loop on successful checking of the condition.

If I was doing this in Golang I would use https://pkg.go.dev/github.com/stretchr/testify/assert#Eventually but I've found a `for` loop in rust with a sleep at the end of it much easier since Rust closures and ownership are trickier to manage - and I've not been able to find a better way.

Reply:
> My suggestion here (if possible) would be to create a looping poll to check a condition (or run the whole test), with a sleep at the end of each loop, and break the loop on successful checking of the condition.

My preference would be that we return a signalling primitive. The problem isn't solvable with a looping poll because the issue isn't "the packet recv timeout expires before the system is ready", the problem is "the test sends the packet before the the socket is listening, so the packet is lost and it will never pass."

Reply:
I've definitely done tests (in other systems) where it's basically

* set check = false
* try 10 times
  * send packet to endpoint
  * timeout after 500ms on getting a packet
  * If successfully received, set check = true, and break
  * If not successful, go around the loop again.

So you get a polling operation that eventually passes, and generally isn't prone to hitting race conditions with a single sleep operation.

Reply:
My worry about that loop, is that it could mask some bugs, for the sake of passing race conditions, for example, it would be indistinguishable from a bug in the proxy that was dropping packets.

Reply:
Ah yep, I do see your point there. That's fair.

> My preference would be that we return a signalling primitive that we await, so the moment it's ready the test starts. 

That makes a lot of sense then - and shouldn't be impossible to do.

Are we thinking we keep the sleeps for now, and then adjust at a later date (or as we track potential flakiness in tests, and fix as needed?) - that would work for me.

Reply:
Yeah I think keeping the sleeps for now is best, because it will require changing a lot more and it's not the worst right now, I've already implemented part of it in this PR, but there's a lot more code for updating our test framework.

Reply:
SGTM!

Did you want to merge this now, or wait until after performance tests are done?